File: apps/image-api/src/services/s3.ts
import {
  CreateBucketCommand,
  DeleteObjectCommand,
  HeadBucketCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { config } from "../config";

const s3 = new S3Client({
  region: config.aws.region,
  endpoint: config.aws.endpoint,
  forcePathStyle: Boolean(config.aws.endpoint),
  credentials: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey,
  },
});

export function getCdnUrl(key: string): string {
  const base = config.cloudfrontUrl.replace(/\/$/, "");
  return `${base}/${key}`;
}

export async function ensureBucket(): Promise<void> {
  try {
    await s3.send(new HeadBucketCommand({ Bucket: config.aws.bucket }));
  } catch {
    await s3.send(new CreateBucketCommand({ Bucket: config.aws.bucket }));
  }
}

export async function uploadObject(
  key: string,
  body: Buffer,
  contentType: string,
): Promise<void> {
  await s3.send(
    new PutObjectCommand({
      Bucket: config.aws.bucket,
      Key: key,
      Body: body,
      ContentType: contentType,
      CacheControl: "public, max-age=31536000, immutable",
    }),
  );
}

export async function deleteObjects(keys: string[]): Promise<void> {
  await Promise.all(
    keys.map((key) =>
      s3.send(
        new DeleteObjectCommand({ Bucket: config.aws.bucket, Key: key }),
      ),
    ),
  );
}

export { s3 };

File: app/services/s3.py
"""S3 (LocalStack in dev) object storage, mirroring services/s3.ts."""

import boto3
from botocore.config import Config as BotoConfig

from config import config

_s3 = boto3.client(
    "s3",
    region_name=config.AWS_REGION,
    endpoint_url=config.AWS_ENDPOINT,
    aws_access_key_id=config.AWS_ACCESS_KEY_ID,
    aws_secret_access_key=config.AWS_SECRET_ACCESS_KEY,
    config=BotoConfig(s3={"addressing_style": "path"} if config.AWS_ENDPOINT else {}),
)


def get_cdn_url(key: str) -> str:
    return f"{config.CLOUDFRONT_URL.rstrip('/')}/{key}"


def ensure_bucket() -> None:
    try:
        _s3.head_bucket(Bucket=config.S3_BUCKET)
    except Exception:
        _s3.create_bucket(Bucket=config.S3_BUCKET)


def upload_object(key: str, body: bytes, content_type: str) -> None:
    _s3.put_object(
        Bucket=config.S3_BUCKET,
        Key=key,
        Body=body,
        ContentType=content_type,
        CacheControl="public, max-age=31536000, immutable",
    )


def delete_objects(keys: list[str]) -> None:
    for key in keys:
        _s3.delete_object(Bucket=config.S3_BUCKET, Key=key)
